aztk/node_scripts/submit.py

```diff
@@ -178,20 +153,98 @@ def upload_error_log(error, application_file_path):
+    return yaml.load(yaml_serialized_task)
+
+
+def ssh_submit(task_sas_url):
+    task = download_task_definition(task_sas_url)
+    download_task_resource_files(task.id, task.resource_files)
+
+    application = load_application(os.path.join(os.environ["AZ_BATCH_TASK_WORKING_DIR"], "application.yaml"))
+
+    cmd = __app_submit_cmd(application)
+    return_code = subprocess.call(cmd.to_str(), shell=True)
+
+    upload_log(config.blob_client, application)
+
+    return return_code
+
+
+def http_request_wrapper(func, *args, timeout=None, max_execution_time=300, **kwargs):
+    start_time = time.clock()
+    while True:
+        try:
+            response = func(*args, timeout=timeout, **kwargs)
+            response.raise_for_status()
+            return response
+        except requests.Timeout:
+            pass
+
+        if (time.clock() - start_time > max_execution_time):
+            raise error.AztkError("Waited {} seconds for request {}, exceeded max_execution_time={}".format(
+                time.clock() - start_time,
+                func.__name__,
+                max_execution_time,
+            ))
+
+
+def _download_resource_file(task_id, resource_file):
+    # timeout = 30 # set to default blob download timeout
+    response = http_request_wrapper(requests.get, url=resource_file.blob_source, timeout=None, stream=True)
+    if resource_file.file_path:
+        write_path = os.path.join(os.environ.get("AZ_BATCH_TASK_WORKING_DIR"), resource_file.file_path)
+        with open(write_path, 'wb') as stream:
+            for chunk in response.iter_content(chunk_size=16777216):
+                stream.write(chunk)
+            return None
+
+    raise error.AztkError("ResourceFile file_path not set.")
+
+
+def download_task_resource_files(task_id, resource_files):
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        futures = {
+            executor.submit(_download_resource_file, task_id, resource_file): resource_file
+            for resource_file in resource_files
+        }
+    done, not_done = concurrent.futures.wait(futures)
+    if not_done:
+        raise error.AztkError("Not all futures completed. {}".format(not_done.pop().result()))
+    errors = [result.result() for result in done if isinstance(result.result(), Exception)]
+    if errors:
+        raise error.AztkError(errors)
+    else:
+        return [result.result() for result in done]
+
+
 if __name__ == "__main__":
     return_code = 1
-    try:
-        return_code = receive_submit_request(os.path.join(os.environ["AZ_BATCH_TASK_WORKING_DIR"], "application.yaml"))
-    except Exception as e:
-        upload_error_log(str(e), os.path.join(os.environ["AZ_BATCH_TASK_WORKING_DIR"], "application.yaml"))
 
-    # force batch task exit code to match spark exit code
-    sys.exit(return_code)
+    if len(sys.argv) == 2:
+        serialized_task_sas_url = sys.argv[1]
+
+        try:
+            return_code = ssh_submit(serialized_task_sas_url)
```

Comment:
what happens if the ssh connection is killed?

Reply:
the ssh session only exists to kick off this process, it does not live until completion. So this return_code is not sent back to the client.

This could be written to storage for retrieval, though. Spark application status, execution time, and return value are still open issues here.

Reply:
ok sounds good then